exporter/exporterhelper/batch_sender.go

```diff
@@ -0,0 +1,461 @@
+// be 0. The optionalReq argument can be nil, make sure to check it before using. maxItems argument is guaranteed to be
+// greater than 0. Context will be propagated from the original request.
+// This API is at the early stage of development and may change without backward compatibility
+// until https://github.com/open-telemetry/opentelemetry-collector/issues/8122 is resolved.
+type BatchMergeSplitFunc func(ctx context.Context, optionalReq Request, req Request, maxItems int) ([]Request, error)
+
+// IdentifyBatchFunc returns an identifier for a Request batch. This function can be used to separate particular
+// Requests into different batches. Batches with different identifiers will not be merged together.
+// Provided context can be used to extract information from the context and use it as a part of the identifier as well.
+// This function is optional. If not provided, all Requests will be batched together.
+// This API is at the early stage of development and may change without backward compatibility
+// until https://github.com/open-telemetry/opentelemetry-collector/issues/8122 is resolved.
+type IdentifyBatchFunc func(ctx context.Context, r Request) string
+
+type shardGetter interface {
+	shard(*request) (*shard, error)
+}
+
+type Batcher struct {
+	cfg            MergeSplitBatcherConfig
+	mergeFunc      BatchMergeFunc
+	mergeSplitFunc BatchMergeSplitFunc
+
+	shardGetter
+
+	shutdownCh chan struct{}
+	goroutines sync.WaitGroup
+	logger     *zap.Logger
+
+	export func(*request) error
+}
+
+// shutdown is invoked during service shutdown.
+func (b *Batcher) shutdown() {
+	close(b.shutdownCh)
+
+	// Wait until all goroutines are done.
+	b.goroutines.Wait()
+}
+
+// newShard gets or creates a batcher corresponding with attrs.
+func (b *Batcher) newShard() *shard {
+	s := &shard{
+		batcher:    b,
+		newRequest: make(chan *request, runtime.NumCPU()),
+	}
+	b.goroutines.Add(1)
+	go s.start()
+	return s
+}
+
+func (b *Batcher) mergeRequests(req1 *request, req2 *request) (*request, error) {
+	r, err := b.mergeFunc(req2.Context(), req1.Request, req2.Request)
+	if err != nil {
+		return nil, err
+	}
+	return &request{
+		baseRequest: baseRequest{ctx: req2.Context()},
+		Request:     r,
+	}, nil
+}
+
+func (b *Batcher) splitMergeRequest(optReq *request, req *request, maxItems int) ([]*request, error) {
+	var optionalReq Request
+	if optReq != nil {
+		optionalReq = optReq.Request
+	}
+	rs, err := b.mergeSplitFunc(req.Context(), optionalReq, req.Request, maxItems)
+	if err != nil {
+		return nil, err
+	}
+	if len(rs) == 0 {
+		return nil, errors.New("empty requests slice returned")
+
+	}
+
+	reqs := make([]*request, 0, len(rs))
+	for _, r := range rs {
+		reqs = append(reqs, &request{
+			baseRequest: baseRequest{ctx: req.Context()},
+			Request:     r,
+		})
+	}
+	return reqs, nil
+}
+
+// NewMergeBatcher creates a new Batcher that merges requests.
+func NewMergeBatcher(cfg MergeBatcherConfig, mf BatchMergeFunc, opts ...BatcherOption) *Batcher {
```

Comment:
Do we need `NewMergeBatcher` and the `Batcher` struct to be public? Why not `WithBatcher` accepts everything that `NewMergeBatcher` accepts?

Reply:
The idea is to provide 2 options for the exporter authors:
- merge only batching 
- merge+split batching

Probably we can move `split` option to another `BatcherOption` and keep one `WithBatcher`

Reply:
Applied splitting as an option